app/src/main/java/me/hufman/androidautoidrive/carapp/music/views/EnqueuedView.kt

```diff
@@ -1,77 +1,219 @@
+				title = title.substring(0, 20) + "..."
+			}
+			val artist = UnicodeCleaner.clean(item.artist ?: "")
+			val songMetaDataText = "${title}\n${artist}"
+
+			return arrayOf(
+					checkmark,
+					coverArtImage,
+					"",
+					songMetaDataText
+			)
 		}
 	}
 
 	init {
 		state as RHMIState.PlainState
 
 		listComponent = state.componentsList.filterIsInstance<RHMIComponent.List>().first()
+		queueImageComponent = state.componentsList.filterIsInstance<RHMIComponent.Image>().first()
+		titleLabelComponent = state.componentsList.filterIsInstance<RHMIComponent.Label>()[0]
+		subtitleLabelComponent = state.componentsList.filterIsInstance<RHMIComponent.Label>()[1]
 
 		songsEmptyList.addRow(arrayOf("", "", L.MUSIC_QUEUE_EMPTY))
 	}
 
 	fun initWidgets(playbackView: PlaybackView) {
-		state.getTextModel()?.asRaDataModel()?.value = L.MUSIC_QUEUE_TITLE
+		// this is required for pagination system
+		listComponent.setProperty(RHMIProperty.PropertyId.VALID, false)
+
 		listComponent.setVisible(true)
-		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,50,*")
+		listComponent.setProperty(RHMIProperty.PropertyId.LIST_COLUMNWIDTH, "57,90,10,*")
 		listComponent.getAction()?.asHMIAction()?.getTargetModel()?.asRaIntModel()?.value = playbackView.state.id
 		listComponent.getAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onClick(it) }
+		listComponent.getSelectAction()?.asRAAction()?.rhmiActionCallback = RHMIActionListCallback { onSelectAction(it) }
 	}
 
 	fun show() {
 		currentSong = musicController.getMetadata()
+		val newQueueMetadata = musicController.getQueue()
+
+		// same queue as before, just select currently playing song
+		if (queueMetadata != null && queueMetadata == newQueueMetadata) {
+			showCurrentlyPlayingSong()
+			return
+		}
+
+		queueMetadata = newQueueMetadata
 		songsList.clear()
-		val songs = musicController.getQueue()
+		val songs = queueMetadata?.songs
 		if (songs?.isNotEmpty() == true) {
 			listComponent.setEnabled(true)
 			listComponent.setSelectable(true)
 			songsList.addAll(songs)
-			listComponent.getModel()?.setValue(songsListAdapter, 0, songsListAdapter.height, songsListAdapter.height)
-
-			// set the selection to the current song
-			val index = songs.indexOfFirst { it.queueId == currentSong?.queueId }
-			if (index >= 0) {
-				state.app.events.values.firstOrNull { it is RHMIEvent.FocusEvent }?.triggerEvent(
-						mapOf(0.toByte() to listComponent.id, 41.toByte() to index)
-				)
+
+			listComponent.requestDataCallback = RequestDataCallback { startIndex, numRows ->
+				showList(startIndex, numRows)
+
+				val endIndex = if (startIndex + numRows >= songsList.size) songsList.size - 1 else startIndex + numRows
+				visibleRows = songsListAdapter.realData.subList(startIndex, endIndex + 1)
+				visibleRowsOriginalMusicMetadata.clear()
+				visibleRows.forEach { musicMetadata ->
+					visibleRowsOriginalMusicMetadata.add(MusicMetadata.copy(musicMetadata))
```

Comment:
There is a chance (and in fact, I encountered it) that this `visibleRowsOriginalMusicMetadata` update can happen during a refresh, which caused the handler thread to crash with a ConcurrentModificationException. One solution might be for `redraw()` to create a clone before it iterates through it, but a safer fix is to create a new local list here with the MusicMetadata objects and then set visibleRowsOriginalMusicMetadata to point to this new list, which I believe is atomic and safe from ConcurrentModificationExceptions. This can be enforced by declaring `visibleRowsOriginalMusicMetadata` to be a List, which is readonly in Kotlin.

In fact, the syntax for this change is very elegant: `visibleRowsOriginalMusicMetadata = visibleRows.map { MusicMetadata.copy(it) }`

Reply:
Oh thanks for pointing that out, surprisingly I hadn't seen this happen yet when I was testing. I updated the visibleRowsOriginalMusicMetadata to be a List that is assigned via the map method and also changed the visibleRows method to use a copy of the sublist rather than the list segment itself.